noir-projects/aztec-nr/aztec/src/note/utils.nr

```diff
@@ -35,48 +37,77 @@ pub fn compute_note_hash_for_read_request<Note, let N: u32, let M: u32>(note: No
     }
 }
 
-pub fn compute_note_hash_for_consumption<Note, let N: u32, let M: u32>(note: Note) -> Field where Note: NoteInterface<N, M> {
+// TODO(#7775): make this not impossible to understand
+pub fn compute_note_hash_for_consumption_internal<Note, let N: u32, let M: u32>(
+    note: Note,
+    note_hash_for_read_request: Field
+) -> Field where Note: NoteInterface<N, M> {
     let header = note.get_header();
-    // There are 3 cases for reading a note intended for consumption:
-    // 1. The note was inserted in this transaction, is revertible, or is not nullified by a revertible nullifier in 
-    //    the same transaction: (note_hash_counter != 0) & (nonce == 0)
-    // 2. The note was inserted in this transaction, is non-revertible, and is nullified by a revertible nullifier in 
-    //    the same transaction: (note_hash_counter != 0) & (nonce != 0)
-    // 3. The note was inserted in a previous transaction: (note_hash_counter == 0) & (nonce != 0)
 
-    // TODO(#7771): inject compute_note_hash(...) func to notes with macros.
-    let note_hash = note.compute_note_hiding_point().x;
-
-    if header.nonce == 0 {
-        // Case 1.
-        // If a note is transient, we just read the note_hash (kernel will hash it with nonce and silo by contract address).
-        note_hash
+    if header.note_hash_counter != 0 {
```

Comment:
The logic here would be significantly less twisted if we decided to drop [this property](https://github.com/AztecProtocol/aztec-packages/blob/6c6f2bbd6c7d129c9b4e8c6028b110e8676bba4d/yarn-project/end-to-end/src/e2e_keys.test.ts#L53) - then we would not need to silo the note hash here (this would also save some gates).

I think this should be considered because I feel the usecase has too many limitations (see the comment in the code there) to really be practical.

Reply:
I went over the comments in that test a bunch of times but am afraid I still don't understand what property they refer to. I'm not even sure if they're talking about address siloing or nonce siloing. So it may be that what I write below is the same as what you're thinking, but I'll leave it here anyway in case it's useful to reference in the future.

I discussed this a little bit with @LeilaWang, and we arrived at a couple of useful conclusions. The real reason why we need multiple kinds of nullif hashes is because the kernel needs a 'hint' as to whether the new nullifier is going to result in the note + nullif being squashed or not. There are multiple reasons why squashing might not happen (e.g. non revertible notes), but there's a single scenario in which we can squash: a revertible transient note (i.e. zero nonce, non-zero note hash counter).

It seems additionally like there is also no need to do contract-siloing of the nullifier - if desired the kernel could always add this later. I'm not sure if this is the kind of siloing you were referring to in this comment?

Given the above, and if we can drop the address siloing, then all we need to do is identify the odd case in which the note is transient (with a non-zero hash counter), and so the read request does not include the nonce, but also non-revertible (the nonce is also non-zero!), and only in this case extend the note hash for read request with the nonce. I'm not sure if this also requires knowing if the nullifier is revertible or not.

Reply:
> I'm not sure if this is the kind of siloing you were referring to in this comment?

Yes, we have a terminology issue here as some people use "siloing" to refer to hashing a hash with a contract address and other people use it to refer to hashing of a hash with whatever else (e.g. siloing with a nonce).

As @iAmMichaelConnor originally coined it it used to mean hashing of hash with contract address and that's what should be used if we don't want the naming to be 💩 (otherwise what is siloed_note_hash?)


> additionally like there is also no need to do contract-siloing of the nullifier
I was referring to the siloing of the note_hash when computing note_hash_for_nullifiy [here](https://github.com/AztecProtocol/aztec-packages/blob/de124e1f9e8f9a3408f1249603ff3eaeb04db724/noir-projects/aztec-nr/aztec/src/note/utils.nr#L95). The valuable property of this is that when we do the siloing there then a note nullifier can be derived just by having the nsk_app and observing note_hashes inserted into the note_hash_tree. This allows a 3rd party to detect that a note has been nullified just by having the nsk_app and watching for new. nullifiers being inserted. See [this function](https://github.com/AztecProtocol/aztec-packages/blob/de124e1f9e8f9a3408f1249603ff3eaeb04db724/yarn-project/end-to-end/src/e2e_keys.test.ts#L87) for how exactly this works. The important property is that the note hash as it is inserted to a note hash tree is part of the preimage of the nullifier as it allows for brute-forcing the nullifiers from public info.

If we decided to drop this property we could simplify the `compute_note_hash_for_nullify` func. Which as I pointed out might be worthwhile because IDK if the "nullifier detection property" is valuable enough given that there are case which are hard to detect (like e.g. the [DeFi withdrawing case](https://github.com/AztecProtocol/aztec-packages/blob/de124e1f9e8f9a3408f1249603ff3eaeb04db724/yarn-project/end-to-end/src/e2e_keys.test.ts#L66)) and it costs gates.

But anyway it feels like we should keep the property for now as we seem to be doing fine with the gate counts.